Added Common++ unit test project based on Gtest.

Dimi1010 opened the pull request:
## Summary
This PR introduces dedicated unit tests for the Common++ module based on the GoogleTest framework as well as several minor improvements. The dependency is resolved via `find_package` and conditionally via `FetchContent` if `find_package` fails.

## What's Changed?
### New dependency - GoogleTest
Added GoogleTest as dependency to the project when `PCAPPP_BUILD_TESTS` is set to `ON`. The dependency resolution happens in one of two ways: `find_package` or `FetchContent`. The project prioritizes the result of `find_package`, to allow environments to supply their own.

#### Restricted FetchContent dependency resolution.
For environments that disallow or discourage the use of `FetchContent` (e.g. brew, vcpkg, etc) an additional project option `PCAPPP_ALLOW_FETCH_CONTENT` (default: `ON`) is added. Setting the option to `OFF` would disable the automatic fetching and require all dependencies that are optionally supplied by fetch content be resolvable by `find_package` in the build environment.

**NB:** The `googletest` dependency is still declared inside the `FetchContent` infrastructure, but the option disables calling `FetchContent_MakeAvailable` effectively disabling the fetch operation.

### New Unit tests
- A new project `Test/Common++` has been added with unit tests.
  - Tested Classes
    - `IPAddress`, `IPv4Address`, `IPv6Address` - tests provided in `IPAddressTests.cpp`
    - `IPNetwork`, `IPv4Network`, `IPv6Network` - tests provided in `IPAddressTests.cpp`
    - `MacAddress` - tests provided in `MacAddressTests.cpp`
    - `PointereVector<T>` - tests provided in `PointerVectorTests.cpp`
    - `LRUList<T>` - tests provided in `LRUListTests.cpp`
  - Tested Functions
    - `byteArrayToHexString` - tests provided in `GeneralUtilsTests.cpp`
    - `hexStringToByteArray` - tests provided in `GeneralUtilsTests.cpp`
    - `cross_platform_memmem` - tests provided in `GeneralUtilsTests.cpp`
    - `align` - tests provided in `GeneralUtilsTests.cpp`
- Added machinery to fetch `googletest` (v1.16.0) during the cmake build process.
- Enabled `googletest` macro definitions for pre-commit's `cppcheck` step.
- Added listener that can be used to extend a unit test to check for memory leaks.

### Fixes and additions
- Added `#include <array>` to `MacAddress.h` to fix transitive include.

Dimi1010 committed:
Added Common++ unit test project based on Gtest.

Dimi1010 committed:
Fixed includes.

Dimi1010 committed:
Changed gtest version to 1.12 which is the last to support Cpp11.

Dimi1010 committed:
Added Gmock in addition to Gtest.

Dimi1010 committed:
Added IPv6 tests.

Dimi1010 committed:
Added IPAddress tests.

Dimi1010 committed:
Lint

Dimi1010 committed:
Fixed matcher error.

Dimi1010 committed:
Added _ipv4 and _ipv6 string literals.

Dimi1010 committed:
Added IPv4 Network unit tests.

Dimi1010 committed:
Added IPv6 Network unit tests.

Dimi1010 committed:
Added IPNetwork unit tests.

tigercosmos commented on Tests/Common++Test/CMakeLists.txt:
```diff
@@ -5,7 +5,7 @@ include(FetchContent)
     googletest
     GIT_REPOSITORY https://github.com/google/googletest.git
-    GIT_TAG release-1.11.0
+    GIT_TAG release-1.12.0
```
I don't know that we already have gtest.

clementperon replied:
We didn't the fetch content will pull it from internet

tigercosmos replied:
I know. I meant I am suprised that there is gtest fetching in current master because we are not using gtest now.

Dimi1010 replied:
> I am suprised that there is gtest fetching in current master

Wdym, @tigercosmos ? There isn't one in the master branch?

clementperon replied:
@tigercosmos you're looking at a specific commit and not the whole MR

Dimi1010 committed:
Added more unit tests.

- Added IPAddress multicast unit tests.
- Added IPNetwork includes unit tests.
- Formatting

Dimi1010 committed:
Added IP(v4/v6)Address match network tests.

Dimi1010 committed:
Lint

Dimi1010 committed:
Added MacAddress constructor from std::array.

Dimi1010 committed:
Added MacAddress unit tests.

Dimi1010 committed:
Added MacAddress tests for output to stream.

Dimi1010 committed:
Added unit tests for PointerVector and test fixture with built-in memleak detection.

Dimi1010 committed:
Lint

Dimi1010 committed:
Included 'googletest' configuration for cppcheck.

Dimi1010 committed:
Fixed include of MemoryLeakDetectorFixture... hopefully the correct way.

Dimi1010 committed:
Fixed wrong variable assert.

Dimi1010 committed:
Added unit test for functions in the general utilities header.

Dimi1010 committed:
Added unit tests for LRUList.

Dimi1010 committed:
Lint

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 committed:
Added cast to ptr type because direct nullptr is deleted function.

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

Dimi1010 committed:
Lint

Dimi1010 committed:
Fixed issue with temporary.

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

Dimi1010 committed:
Lint - hopefully.

Dimi1010 committed:
Pulled operator << overloads into the 'pcpp' namespace to leverage ADL as it does not consider functions in the global namespace.

- Pulled operator<< for IpAddress into the `pcpp` namespace.
- Pulled operator<< for IpV4Address into the `pcpp` namespace.
- Pulled operator<< for IpV6Address into the `pcpp` namespace.
- Pulled operator<< for IpNetwork into the `pcpp` namespace.
- Pulled operator<< for IpV4Network into the `pcpp` namespace.
- Pulled operator<< for IpV6Network into the `pcpp` namespace.
- Pulled operator<< for MacAddress into the `pcpp` namespace.

Dimi1010 committed:
Renamed MacAddress ToStream test to be consistent with IPAddress tests of the same nature.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 committed:
CMakeLists lint...

Dimi1010 commented:
So far, it hasn't been integrated into the workflow or the codecov as I'm not really sure how precisely that is achieved.

seladb commented:
@Dimi1010 @tigercosmos @egecetin @clementperon let's discuss this change: if we want to move away from the existing tests infra I think it should be a consternated effort to migrate all tests to Gtest. It'll take time, but may be feasible if all of us are working on it together. The question is: can we all commit to work on it before the next version?

egecetin commented:
@seladb What is the expected date of the new release? If it will be about 25.04 it is easy to work on it by planning. The more important question is, we talked about the gtest in #1463 to improve performance. Is everyone thinking moving Gtest is better, or anyone have any doubts (for performance, maintainability or maybe interface related)? It is better to share it if have any, before starting to work.

In my opinion, it will reduce maintaining requirements because of gtest has lots of developers and they already fix their bugs and has a good interface for a C++ code and have integration with the external tools/extensions but for performance I don't expect a drastically improvement. So, I think moving to gtest is feasible.

seladb commented:
> @seladb What is the expected date of the new release? If it will be about 25.04 it is easy to work on it by planning. The more important question is, we talked about the gtest in #1463 to improve performance. Is everyone thinking moving Gtest is better, or anyone have any doubts (for performance, maintainability or maybe interface related)? It is better to share it if have any, before starting to work.
> 
> In my opinion, it will reduce maintaining requirements because of gtest has lots of developers and they already fix their bugs and has a good interface for a C++ code and have integration with the external tools/extensions but for performance I don't expect a drastically improvement. So, I think moving to gtest is feasible.

I hope we can release faster, hopefully before 25.04. To be honest, I don't think performance is a good enough reason to switch to Gtest. The main reason would be having more standardize code and less custom-made stuff. But as I mentioned, it will require a lot of work (and most of it is boring - rewrite tests...), so the question is - can we commit to that and do we think it's worth the effort?

Dimi1010 commented:
I think it will be beneficial overall. I think the rollout can be done in phases (project by project) similar to the way we did the clang-format roll out. I could probably work through it slowly as I get free time.

Also, a lot of the current tests could probably benefit by the expanded non-critical failure support, I know we have something similar in the PTF framework but its not exactly the same. In GTest a non-critical failure (EXPECT_*) the test is still marked as failed, but continues to run to catch other errors in the unit test if they exist. Fatal failures (ASSERT_*) are mostly used for conditions that invalidate the rest of the unit test. The fact that the current framework only catches 1 error per unit test has been somewhat annoying to me.

seladb commented:
> I think it will be beneficial overall. The rollout can be done in phases (project by project) similar to the way we did the clang-format roll out. I could probably work through it slowly as I get free time.

The difference between this and clang-format is that this project will be much slower and more tedious because we need to go over all the tests and port them one by one. Maybe we can open a separate branch for it so we don't merge unfinished work to `dev` and block `master`?

Anyway, this will require more time and commitment from everyone, and I wonder if we can/want to do it.

However, if we open a separate branch and it doesn't block anything then the risk is quite minimal 🤔 


> Also, a lot of the current tests could probably benefit by the expanded non-critical failure support, I know we have something similar in the PTF framework but its not exactly the same. In GTest a non-critical failure (EXPECT_*) the test is still marked as failed, but continues to run to catch other errors in the unit test if they exist. Fatal failures (ASSERT_*) are mostly used for conditions that invalidate the rest of the unit test. The fact that the current framework only catches 1 error per unit test has been somewhat annoying to me.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

# Conflicts:
#	.pre-commit-config.yaml

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

# Conflicts:
#	Common++/header/IpAddress.h
#	Common++/header/MacAddress.h

Dimi1010 committed:
Split common unit tests into smaller tests.

Dimi1010 committed:
Added const qualifier to MacAddress::Zero.

Dimi1010 committed:
Lint

Dimi1010 committed:
Added basic single threaded logger unit tests.

Dimi1010 committed:
Simplified logger fixture setup and teardown.

Dimi1010 committed:
Added explicit main file.

Dimi1010 committed:
Added conditional skipping of memory leak checks for MSVC Release builds.

Dimi1010 committed:
Added build information to the test executable.

Dimi1010 committed:
Added tests for CoreMask transformations.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

# Conflicts:
#	Common++/header/MacAddress.h

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 committed:
Removed make_unique usage.

Dimi1010 committed:
Lint

Dimi1010 committed:
Added const to MacAddress::Broadcast static variable.

Dimi1010 committed:
Added Broadcast unit test.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

# Conflicts:
#	Tests/CMakeLists.txt

Dimi1010 committed:
Bump googletest to v1.16.0

Dimi1010 committed:
Added unit tests for DynamicObjectPool

Dimi1010 committed:
cmake formatting.

Dimi1010 committed:
Include cleanup

Dimi1010 committed:
Added precompiled header

Dimi1010 committed:
Added unit tests for timespec to timeval conversions.

Dimi1010 committed:
Updated logger tests.

Dimi1010 committed:
Lint

Dimi1010 committed:
Version print formatting.

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

Dimi1010 committed:
Added todo for googletest fetch content.

Dimi1010 committed:
Added common unit test to run_tests CI script.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 committed:
Lint

Dimi1010 committed:
Updated packet only to run common + packet suites.

Dimi1010 committed:
Fixed explicit ctor

Dimi1010 committed:
Updated run_tests_windows to have separate functions for calls to subprocess.

Dimi1010 committed:
Added common test to windows CI.

Dimi1010 committed:
Lint

Dimi1010 committed:
Delayed startup of tcpreplay until pcap++ tests.

Dimi1010 committed:
Added tests for new functions.

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

Dimi1010 committed:
Fixed memory leak fixture not respecting mem leak checks.

Dimi1010 committed:
Changed MemoryLeak Fixture to default fixture under TEST macro.

Dimi1010 committed:
Temporarely disabled memory leak fixture testing due to false positives from google test.

Dimi1010 committed:
Changed MemoryLeakFixture into a gtest listener.

Dimi1010 committed:
Updated log tests to utilize functor object instead of static function.

Dimi1010 committed:
Reverted on test end back to free all memory.

Dimi1010 committed:
Fixed visibility.

Dimi1010 committed:
Lint

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 committed:
Fix cppcheck warnings.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 commented:
@tigercosmos In my tests, it didn't work having the memory tracking encompass all the tests as a single start/stop pair. 

That is because gtest appears to allocate on heap from thread local variables during the tests. That memory isn't released until the main thread actually exits when the program shuts down, and gets detected as a false positive. :/

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

seladb commented on Tests/Common++Test/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+include(FetchContent)
+
+# TODO: This may have issues with brew as it doesn't allow the use of FetchContent.
+FetchContent_Declare(googletest GIT_REPOSITORY https://github.com/google/googletest.git GIT_TAG v1.16.0)
```
As the `TODO` comment mentions, package managers don't approve packages that use `FetchContent_Declare`, we already had this experience before...

Dimi1010 replied:
I did start another branch [[1]](https://github.com/Dimi1010/PcapPlusPlus/tree/feature/gtest-framework-packet%2B%2B) off of this one, for the Packet++ tests, to see how it would go. You can take a look, if you want. The new project is `Tests/Packet++Test-google`.

It actually mostly involves trivial changes, like macro replacements.
- `PTF_TEST_CASE` -> `TEST(Fixture, TestName)`
- `PTF_ASSERT_EQUAL` -> `ASSERT_EQ` / `EXPECT_EQ`
- `PTF_ASSERT_NULL` -> `ASSERT_EQ(statement, nullptr)`
- `PTF_ASSERT_NOT_NULL` -> `ASSERT_NE(statement, nullptr)`

The only non-trivial assertion so far has been `PTF_ASSERT_BUF_COMPARE` which was replaced by `EXPECT_TRUE(test::BuffersMatch(actualBuf, actualBufSize, expectedBuf, expectedBufSize))` which ended up being a more comprehensive check IMO. [[2]](https://github.com/Dimi1010/PcapPlusPlus/blob/feature/gtest-framework-packet%2B%2B/Tests/Packet%2B%2BTest-google/Utils/AssertionExtensions.cpp#L31-L94).

The loading utilities needed a revamp, as I didn't want to use macros everywhere and the current ones had a limitation of always using the current working directory (has issues with out of source builds).

For the majority of the cases the code: 
```cpp
timeval time;
gettimeofday(&time, nullptr);

READ_FILE_AND_CREATE_PACKET(1, "PacketExamples/TcpPacketNoOptions.dat");
// rest of test case
```
ended up being a one liner:
```cpp
// std::unique_ptr<RawPacket>
auto rawPacket1 = test::createPacketFromHexResource("PacketExamples/TcpPacketWithOptions3.dat", /* optional packet factory */, /* optional data loader */);
```

Dimi1010 replied:
That is why the Packet++ port is in a different target. It will allow incremental PRs as tests are ported.

A way I found to diff the changes between the tests is to diff the `Packet++Test/Tests/testname.cpp` and `Packet++Test-google/Tests/testname.cpp`. Don't think it can be done on github's PR interface, so it will probably require pulling to local, but it gives quite a clean diff.

<img width="2026" height="1202" alt="image" src="https://github.com/user-attachments/assets/4fcb0557-0bd2-4fa8-af82-836755d1ee2b" />

seladb replied:
I guess we can do it incrementally, file by file (or few files in a PR), if you have the patience to do all this work 🙂

Dimi1010 replied:
I don't mind doing it. I think it will be a worthwhile change.

seladb replied:
Ok sure 👍 
I'll try to review this PR in the coming days

seladb replied:
Actually, even this PR is pretty huge. Maybe we can break it into smaller PRs to make it easier to review?

seladb replied:
Perhaps you're right; I'm not familiar with the requirements of the different package managers. Maybe we can build the library without the tests, which will require changing the recipe in each package. To be on the safe side, I'd prefer using a header-only library if possible

Dimi1010 replied:
Hmm... I missed that they offered amalgamated versions. I guess we can do that?

Upon first glance they mostly appear to be offered for migration compatibility from v2. The maintainers do recommend moving away from those headers in the migration instructions and inside the amalgamated header, as they apparently increase the compilation times significantly. https://github.com/catchorg/Catch2/blob/v3.9.1/docs/migrate-v2-to-v3.md#how-to-migrate-projects-from-v2-to-v3

seladb replied:
Maybe we can re-consider using Gtest, however we need to make sure that:
1. All package managers are ok with using `FetchContent` if it's not used as part of building the package, otherwise we'll do this change for nothing. Any idea how we can verify that?
2. What's the overhead in fetching and compiling Gtest? How much time does it add to building PcapPlusPlus with the tests?

seladb replied:
Ah ok, that makes sense. I just wanted to confirm that we're not missing anything.

So I guess we decided to go with Gtest?

seladb replied:
Yes I think that's a good idea 👍

Dimi1010 replied:
Homebrew has also been updated https://github.com/Homebrew/homebrew-core/pull/233719

Conan needs no update as we already don't build tests there on install.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 committed:
Commented out memory leak warning as it is disabled in all modes.

Dimi1010 committed:
Added googletest 1.16 to 3rdParty dependencies.

Dimi1010 committed:
Added googletest to pre-commit exclusions.

Dimi1010 committed:
Fixup regex string.

Dimi1010 committed:
Disabled clang format for commented line.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

# Conflicts:
#	.pre-commit-config.yaml

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

# Conflicts:
#	Tests/CMakeLists.txt

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 committed:
Revert "Added googletest 1.16 to 3rdParty dependencies."

This reverts commit 789a17eebc4d7349fbe421f3b085929fb495ba29.

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

Dimi1010 committed:
Add logic to prefer resolving gtest via find_package, before conditionally fetching it from FetchContent.

Dimi1010 commented on Tests/Common++Test/Tests/GeneralUtilsTests.cpp:
```diff
@@ -0,0 +1,37 @@
+
+#include "GeneralUtils.h"
+
+namespace pcpp
```
Should probably change the namespace to `pcpp_tests`...

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

Dimi1010 committed:
Merge branch 'dev' into feature/gtest-framework

# Conflicts:
#	.github/workflows/build_and_test.yml
#	ci/run_tests/run_tests.py
#	ci/run_tests/run_tests_windows.py

Dimi1010 committed:
Update build directory.

Dimi1010 committed:
Update tests to pcpp_test namespace.

Dimi1010 committed:
Add common test program to ci script.

Dimi1010 committed:
Wrap OpenCppCoverage in lambda.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/gtest-framework

Dimi1010 committed:
Consolidated LRU tests.

Dimi1010 committed:
Fixup ci script.

Dimi1010 committed:
Lint
